instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.ResourceAttributes;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * An easier alternative to {@link io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider}, which
+ * avoids some common pitfalls and boilerplate.
+ *
+ * <p>An example of how to use this interface can be found in {@link ManifestResourceProvider}.
+ */
+@SuppressWarnings({"unchecked", "rawtypes"})
+public abstract class AttributeResourceProvider<D> implements ConditionalResourceProvider {
+
+  private final AttributeProvider<D> attributeProvider;
+
+  public class AttributeBuilder implements AttributeProvider.Builder<D> {
+
+    private AttributeBuilder() {}
+
+    @CanIgnoreReturnValue
+    @Override
+    public <T> AttributeBuilder add(AttributeKey<T> key, Function<D, Optional<T>> getter) {
+      attributeGetters.put((AttributeKey) key, Objects.requireNonNull((Function) getter));
+      return this;
+    }
+  }
+
+  private static final ThreadLocal<Resource> existingResource = new ThreadLocal<>();
```

Comment:
I think using thread local isn't necessary here, could just use an instance variable.

Reply:
There's one use case where we need it:

1. service name is already present in existing (because another provider has detected it), but version is not
2. should apply will return true (because version is not found)
3. `createResource` does not have access to `existing` - it's not passed as a param

Reply:
I don't get why this would require keeping the existing resource in a thread local, instance variable could be set in the `shouldApply` and used in `createResource` the same way as the thread local. Imo there isn't anything that would require that these classes have to be stateless. If you want you could figure out which attributes need to be set in `shouldApply` and keep that info as instance variable and use it in `createResource` then you probably wouldn't need the existing resource there at all.

Reply:
good idea!

Reply:
Along the way, I also found a bug - so let me try to get that into the release